### PR TITLE
Bug fixes in DataBuffer

### PR DESCRIPTION
**Pull request by balazsracz**
There were several corner cases in appending to a Datauffer that were incorrectly handled and causing crashes in DirectHub.
This PR fixes bugs and adds tests including several fuzz tests for databuffers.

### CODE REVIEW

File: src/utils/DataBuffer.hxx [hunk 1]
```diff
@@ -176,8 +199,13 @@ private:
 
 using DataBufferPtr = std::unique_ptr<DataBuffer, BufferDelete<uint8_t[]>>;
 
+class LinkedDataBufferPtr;
```
Comment (atanisoft) [inline src/utils/DataBuffer.hxx:202]: this declaration seems unnecessary since the class body is just below. Any particular reason this is needed here?

File: src/utils/DataBuffer.hxx [hunk 2]
```diff
@@ -456,37 +531,70 @@ public:
...
+    /// is no shared ownership of this->tail_.
     /// @return true if append succeeded. If false, nothing was changed.
-    bool try_append_from(const LinkedDataBufferPtr &o)
+    bool try_append_from(const LinkedDataBufferPtr &o, bool add_link = false)
     {
         if (!o.size())
         {
             return true; // zero bytes, nothing to do.
         }
+        if (!size_) {
```
Comment (atanisoft) [inline src/utils/DataBuffer.hxx:544]: brace formatting?

**Review by bakerstu — APPROVED** [review]
Please address @atanisoft comments. I agree with them.

File: src/utils/DataBuffer.hxx (hunk 1 above)
Comment (balazsracz) [reply src/utils/DataBuffer.hxx]: removed

File: src/utils/DataBuffer.hxx (hunk 2 above)
Comment (balazsracz) [reply src/utils/DataBuffer.hxx]: fixed all
